File: packages/sdk-php/tests/SecurityTest.php
<?php

declare(strict_types=1);

namespace Syrotp\Sdk\Tests;

use GuzzleHttp\Psr7\Response;
use Syrotp\Sdk\Errors\SyrotpAuthError;
use Syrotp\Sdk\SyrotpClient;
use Syrotp\Sdk\Tests\Support\Fixtures;
use Syrotp\Sdk\Tests\Support\RecordingLogger;
use PHPUnit\Framework\TestCase;

/**
 * Security canary tests. Pin the rules from `docs/sdk-generation.md` §5:
 *
 *   - Never log the API key.
 *   - Never log the request body (which includes the user's phone).
 *   - Never embed credentials in error stringification.
 *
 * These tests use stable canary values that, if they ever appear in
 * captured logs or error renderings, prove a regression.
 */
final class SecurityTest extends TestCase
{
    /** Sentinel chosen to be unmistakable if it ever leaks. */
    private const CANARY_API_KEY = 'sk_live_TESTSENTINEL_DO_NOT_LOG_THIS';
    private const CANARY_PHONE = '+99999999999999';

    // ----- API key MUST NOT appear in error strings ------------------------

    public function testAuthErrorStringDoesNotContainApiKey(): void
    {
        $err = new SyrotpAuthError('unauthorized', 'missing creds', 401);
        self::assertStringNotContainsString(self::CANARY_API_KEY, (string) $err);
    }

    public function testClientDoesNotDefineToStringThatCouldLeakApiKey(): void
    {
        // The SDK deliberately does NOT define a `__toString` on
        // `SyrotpClient`. If a future change adds one it MUST NOT
        // include the api_key — this canary would catch a regression.
        // (`print_r` / `var_dump` will still dump private properties;
        // those are debug tools and should not be wired into log
        // pipelines.)
        self::assertFalse(
            method_exists(SyrotpClient::class, '__toString'),
            'SyrotpClient should not define __toString — risk of leaking apiKey',
        );
    }

    // ----- request body MUST NOT be logged ---------------------------------

    public function testPhoneDoesNotAppearInLoggerOutput(): void
    {
        $logger = new RecordingLogger();
        $client = Fixtures::makeClient(
            queue: [new Response(201, [], json_encode(Fixtures::verificationPayload()))],
            apiKey: self::CANARY_API_KEY,
            logger: $logger,
        );
        $client->startVerification(phone: self::CANARY_PHONE, purpose: 'login');

        foreach ($logger->rendered() as $line) {
            self::assertStringNotContainsString(
                self::CANARY_PHONE,
                $line,
                "the user's phone MUST NOT appear in any syrotp log line; saw: {$line}",
            );
            self::assertStringNotContainsString(
                self::CANARY_API_KEY,
                $line,
                "the api_key MUST NOT appear in any syrotp log line; saw: {$line}",
            );
        }
    }

    // ----- cleartext warning behavior --------------------------------------

    public function testCleartextToPublicHostLogsWarning(): void
    {
        $logger = new RecordingLogger();
        new SyrotpClient(
            baseUrl: 'http://otp.example.com',
            apiKey: 'sk_live_x',
            logger: $logger,
        );
        $matching = array_filter(
            $logger->records,
            static fn (array $r): bool => str_contains($r['rendered'], 'plain HTTP'),
        );
        self::assertNotEmpty($matching, 'expected a plain-HTTP warning for a public host');
    }

    /**
     * @dataProvider privateHosts
     */
    public function testPrivateOrLoopbackHostsDoNotWarn(string $baseUrl): void
    {
        $logger = new RecordingLogger();
        new SyrotpClient(
            baseUrl: $baseUrl,
            apiKey: 'sk_live_x',
            logger: $logger,
        );
        $cleartextWarnings = array_filter(
            $logger->records,
            static fn (array $r): bool => str_contains($r['rendered'], 'plain HTTP'),
        );
        self::assertSame([], $cleartextWarnings, 'private/loopback host should not trigger plain-HTTP warning');
    }

    /** @return array<string, array{0: string}> */
    public static function privateHosts(): array
    {
        return [
            'localhost' => ['http://localhost:3000'],
            'loopback v4' => ['http://127.0.0.1:3000'],
            'rfc1918 10.x' => ['http://10.0.0.1'],
            'rfc1918 192.168.x' => ['http://192.168.1.1'],
            'rfc1918 172.16.x' => ['http://172.16.0.1'],
            'link-local 169.254' => ['http://169.254.1.1'],
        ];
    }
}

File: packages/sdk-kotlin/src/test/kotlin/io/orotp/sdk/SecurityTest.kt
package io.syrotp.sdk

import okhttp3.mockwebserver.MockResponse
import okhttp3.mockwebserver.MockWebServer
import org.junit.jupiter.api.AfterEach
import org.junit.jupiter.api.Assertions.assertFalse
import org.junit.jupiter.api.Assertions.assertTrue
import org.junit.jupiter.api.BeforeEach
import org.junit.jupiter.api.Test
import java.io.ByteArrayOutputStream
import java.util.logging.Level
import java.util.logging.Logger
import java.util.logging.SimpleFormatter
import java.util.logging.StreamHandler

/**
 * Canary tests for `docs/sdk-generation.md` §5.
 *
 * If the api_key or the user's phone ever leak into a log line or an
 * error rendering, these will fail.
 */
class SecurityTest {

    private lateinit var server: MockWebServer

    @BeforeEach fun setUp() { server = MockWebServer().also { it.start() } }
    @AfterEach fun tearDown() { server.shutdown() }

    @Test
    fun `api_key never appears in error toString`() {
        val rendered = SyrotpAuthError("unauthorized", "missing creds", 401).toString()
        assertFalse(rendered.contains(CANARY_API_KEY))
    }

    @Test
    fun `phone never appears in any syrotp logger output`() {
        server.enqueue(
            MockResponse().setResponseCode(201).setBody(
                TestServer.verificationJson(
                    sendTo = "+1",
                    message = "VERIFY ABC",
                ),
            ),
        )

        val captured = captureSyrotpLogs {
            val client = SyrotpClient(
                baseUrl = server.url("/").toString().trimEnd('/'),
                apiKey = CANARY_API_KEY,
                timeoutMs = 5_000,
            )
            client.startVerification(phone = CANARY_PHONE, purpose = "login")
        }
        assertFalse(captured.contains(CANARY_PHONE), "phone leaked into syrotp logger: $captured")
        assertFalse(captured.contains(CANARY_API_KEY), "api_key leaked into syrotp logger: $captured")
    }

    @Test
    fun `cleartext base_url to public host logs warning`() {
        val captured = captureSyrotpLogs {
            // Just construct — no requests needed for the warning.
            SyrotpClient(baseUrl = "http://otp.example.com", apiKey = "sk_live_x")
        }
        assertTrue(captured.contains("plain HTTP"), "expected cleartext warning, got: $captured")
    }

    @Test
    fun `cleartext to localhost or RFC1918 does not warn`() {
        val captured = captureSyrotpLogs {
            SyrotpClient(baseUrl = "http://localhost:3000", apiKey = "sk_live_x")
            SyrotpClient(baseUrl = "http://127.0.0.1:3000", apiKey = "sk_live_x")
            SyrotpClient(baseUrl = "http://10.0.0.1", apiKey = "sk_live_x")
            SyrotpClient(baseUrl = "http://192.168.1.1", apiKey = "sk_live_x")
            SyrotpClient(baseUrl = "http://172.16.0.1", apiKey = "sk_live_x")
        }
        assertFalse(captured.contains("plain HTTP"), "should not warn for private hosts: $captured")
    }

    private fun captureSyrotpLogs(block: () -> Unit): String {
        val logger = Logger.getLogger("syrotp")
        val buf = ByteArrayOutputStream()
        val handler = StreamHandler(buf, SimpleFormatter()).apply { level = Level.ALL }
        val originalLevel = logger.level
        val originalUseParents = logger.useParentHandlers
        logger.addHandler(handler)
        logger.useParentHandlers = false
        logger.level = Level.ALL
        try {
            block()
        } finally {
            handler.flush()
            logger.removeHandler(handler)
            logger.useParentHandlers = originalUseParents
            logger.level = originalLevel
        }
        return buf.toString(Charsets.UTF_8)
    }

    private companion object {
        // Sentinels chosen so they're unmistakable if they leak.
        private const val CANARY_API_KEY = "sk_live_TESTSENTINEL_DO_NOT_LOG_THIS"
        private const val CANARY_PHONE = "+99999999999999"
    }
}
